nixos/power-management: add bootCommands and pre-shutdown hooks

kajusnau opened the pull request:
1. Fixed `powerDownCommands` to also run before shutdown, as the option implies.
2. Added new `bootCommands` option:
   - Runs the specified commands once after boot
   - Runs after `powerUpCommands`
3. Adjusted the provided examples to avoid duplicates.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [x] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SuperSandro2000 commented:
Those 3 changes should be separate commits to make reviewing easier ...

SuperSandro2000 commented on nixos/modules/config/power-management.nix:
```diff
@@ -2,6 +2,12 @@
+    mkIf
+    mkOption
+    types
+    ;
```
I personally do not think that this change improves anything really but if you really want to we can keep it.

SuperSandro2000 commented on nixos/modules/config/power-management.nix:
```diff
@@ -13,26 +19,29 @@ in
+        type = types.lines;
         default = "";
+        example = literalExpression ''
+          "''${pkgs.pulseaudio}/bin/pactl set-sink-volume @DEFAULT_SINK@ 75%"
```
This is a really opinionated example. Is there nothing generic we can write here?

kajusnau replied:
Open to ideas.
How about `"${pkgs.util-linux}/bin/rfkill unblock all"` ?

SuperSandro2000 commented on nixos/modules/config/power-management.nix:
```diff
@@ -41,10 +50,10 @@ in
         default = "";
-        example = lib.literalExpression ''
+        example = literalExpression ''
           "''${pkgs.hdparm}/sbin/hdparm -B 255 /dev/sda"
```
If we replaced the example for powerUpCommands, should we not also replace this one?

kajusnau replied:
I replaced the `powerUpCommands` example simply because it was exactly the same as `powerDownCommands`.
This example is still valid.

kajusnau commented on nixos/modules/config/power-management.nix:
```diff
@@ -70,31 +85,57 @@ in
+        wantedBy = [ "multi-user.target" ];
+        script = ''
+          ${cfg.powerUpCommands}
+          ${cfg.bootCommands}
```
Thanks, agreed and changed.

kajusnau committed:
nixos/power-management: fix powerDownCommands to also run before shutdown

kajusnau commented:
Resolved conflicts and rebased

kajusnau committed:
nixos/power-management: add bootCommands option and adjust examples

- add new bootCommands option:
  - runs the specified commands once after boot
  - runs before powerUpCommands
- adjusted some example commands to avoid duplicates
